pubsub/iterator.go

```diff
@@ -528,39 +518,103 @@ func (it *messageIterator) sendModAck(m map[string]*ackResultWithID, deadline ti
-			if ok {
-				md = apiErr.Metadata()
-				st = apiErr.GRPCStatus()
-			}
-			resultsByAckID := make(map[string]*ackResultWithID)
+			resultsByAckID := make(map[string]*AckResult)
 			for _, ackID := range toSend {
 				resultsByAckID[ackID] = m[ackID]
 			}
-			// TODO(hongalex): retry the ackIDs with transient retriable errors.
-			_, _ = processResults(st, resultsByAckID, md)
+
+			st, md := extractMetadata(err)
+			_, toRetry := processResults(st, resultsByAckID, md)
+			if len(toRetry) > 0 {
+				// Retry modacks/nacks in a separate goroutine.
+				go func() {
+					it.retryModAcks(toRetry, deadlineSec)
+				}()
+			}
 		}
 	}
 }
 
-func (it *messageIterator) sendAckIDRPC(ackResultsByAckID map[string]*AckResult, maxSize int, call func([]string) error) bool {
-	ackIDs := make([]string, 0, len(ackResultsByAckID))
-	for k := range ackResultsByAckID {
-		ackIDs = append(ackIDs, k)
+func (it *messageIterator) retryAcks(m map[string]*AckResult) {
+	ctx, cancel := context.WithTimeout(context.Background(), exactlyOnceDeliveryRetryDeadline)
+	defer cancel()
+	bo := newExactlyOnceBackoff()
+	for {
+		if ctx.Err() != nil {
+			for _, r := range m {
+				ipubsub.SetAckResult(r, AcknowledgeStatusOther, ctx.Err())
+			}
+			return
+		}
+		// Don't need to split map since this is the retry function and
+		// there is already a max of 2500 ackIDs here.
+		ackIDs := make([]string, 0, len(m))
+		for k := range m {
+			ackIDs = append(ackIDs, k)
+		}
+		cctx2, cancel2 := context.WithTimeout(ctx, 60*time.Second)
+		defer cancel2()
+		err := it.subc.Acknowledge(cctx2, &pb.AcknowledgeRequest{
+			Subscription: it.subName,
+			AckIds:       ackIDs,
+		})
+		st, md := extractMetadata(err)
+		_, toRetry := processResults(st, m, md)
+		if len(toRetry) == 0 {
+			return
+		}
+		time.Sleep(bo.Pause())
+		m = toRetry
 	}
-	var toSend []string
-	for len(ackIDs) > 0 {
-		toSend, ackIDs = splitRequestIDs(ackIDs, ackIDBatchSize)
-		if err := call(toSend); err != nil {
-			// The underlying client handles retries, so any error is fatal to the
-			// iterator.
-			it.fail(err)
-			return false
+}
+
+func (it *messageIterator) retryModAcks(m map[string]*AckResult, deadlineSec int32) {
+	bo := newExactlyOnceBackoff()
+	retryCount := 0
+	ctx, cancel := context.WithTimeout(context.Background(), exactlyOnceDeliveryRetryDeadline)
+	defer cancel()
+	for {
+		// If context is done, complete all remaining Nacks with DeadlineExceeded
+		// ModAcks are not exposed to the user so these don't need to be modified.
+		if ctx.Err() != nil {
+			if deadlineSec == 0 {
+				for _, r := range m {
+					ipubsub.SetAckResult(r, AcknowledgeStatusOther, ctx.Err())
+				}
+			}
+			return
+		}
+		// Only retry modack requests up to 3 times.
+		if deadlineSec != 0 && retryCount > 3 {
+			ackIDs := make([]string, 0, len(m))
+			for k := range m {
+				ackIDs = append(ackIDs, k)
+			}
+			log.Printf("automatic lease modack retry failed for following IDs: %v", ackIDs)
```

Comment:
nit: does pubsub log today? I don't think most(any?) of our clients do right now

Reply:
Summarizing our direct conversation, we don't log right now and most Go clients don't. There should be a better way to do logging and we should expose hooks for users to do this.